overlord/hookstate/hooks.go

```diff
@@ -72,6 +76,141 @@ func SetupPreRefreshHook(st *state.State, snapName string) *state.Task {
 	return task
 }
 
+type gateAutoRefreshHookHandler struct {
+	context *Context
+	refreshAppAwareness bool
+}
+
+func (h *gateAutoRefreshHookHandler) Before() error {
+	st := h.context.State()
+	st.Lock()
+	defer st.Unlock()
+
+	tr := config.NewTransaction(st)
+	experimentalRefreshAppAwareness, err := features.Flag(tr, features.RefreshAppAwareness)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if !experimentalRefreshAppAwareness {
+		return nil
+	}
+
+	h.refreshAppAwareness = true
+	if err := runinhibit.LockWithHint(h.context.InstanceName(), runinhibit.HintInhibitedGateRefresh); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (h *gateAutoRefreshHookHandler) Done() error {
+	ctx := h.context
+	st := ctx.State()
+	ctx.Lock()
+	defer ctx.Unlock()
+
+	snapName := h.context.InstanceName()
+
+	var action snapstate.GateAutoRefreshAction
+	a := ctx.Cached("action")
+	if a == nil {
+		// action is unset if the gate-auto-refresh hook exits 0 without
+		// invoking --hold/--proceed; this means proceed (except for respecting
+		// refresh inhibit).
+		if err := snapstate.ProceedWithRefresh(st, snapName); err != nil {
+			return err
+		}
+		action = snapstate.GateAutoRefreshProceed
+	} else {
+		var ok bool
+		action, ok = a.(snapstate.GateAutoRefreshAction)
+		if !ok {
+			return fmt.Errorf("internal error: unexpected action type %T", a)
+		}
+	}
+
+	if !h.refreshAppAwareness {
+		return nil
+	}
+
+	// action is set if snapctl refresh --hold/--proceed was called from the hook.
+	switch action {
+	case snapstate.GateAutoRefreshHold:
+		if err := runinhibit.Unlock(snapName); err != nil {
+			return fmt.Errorf("cannot release inhibit lock of snap %s: %v", snapName, err)
+		}
+	case snapstate.GateAutoRefreshProceed:
+		// we have HintInhibitedGateRefresh lock already when running the hook,
+		// change it to HintInhibitedForRefresh.
+		if err := runinhibit.LockWithHint(snapName, runinhibit.HintInhibitedForRefresh); err != nil {
```

Comment:
I'm not quite sure if this is right if action was nil (default behavior), in that case we want the to do the soft check as usual, in the case of actual proceed if we want to ignore the check so this is more correct